include/nbl/builtin/hlsl/blur/box_blur.hlsl

```diff
@@ -0,0 +1,318 @@
+
+static const uint32_t ITEMS_PER_THREAD = ( AXIS_DIM + WORKGROUP_SIZE - 1 ) / WORKGROUP_SIZE;
+static const uint32_t ITEMS_PER_WG = ITEMS_PER_THREAD * WORKGROUP_SIZE;
+
+static const uint32_t arithmeticSz = nbl::hlsl::workgroup::scratch_size_arithmetic<ITEMS_PER_WG>::value; 
+static const uint32_t broadcastSz = nbl::hlsl::workgroup::scratch_size_broadcast;
+static const uint32_t scratchSz = arithmeticSz + broadcastSz;
+
+groupshared uint32_t scratch[ scratchSz ];
+
+template<typename T, uint16_t offset>
+struct ScratchProxy
+{
+	T get( const uint32_t ix )
+	{
+		return nbl::hlsl::bit_cast< T, uint32_t >( scratch[ ix + offset ] );
+	}
+	void set( const uint32_t ix, NBL_CONST_REF_ARG(T) value )
+	{
+		scratch[ ix + offset ] = nbl::hlsl::bit_cast< uint32_t, T >( value );
+	}
+
+	void workgroupExecutionAndMemoryBarrier()
+	{
+		nbl::hlsl::glsl::barrier();
+	}
+};
+
+ScratchProxy<float32_t, 0> prefixSumsAccessor; 
+ScratchProxy<float32_t, arithmeticSz> broadcastAccessor;
+
+struct prefix_sum_t
+{
+	using bin_op_t = nbl::hlsl::plus<float32_t>;
+	using type_t = bin_op_t::type_t;
+	
+	type_t operator()( NBL_CONST_REF_ARG(type_t) value )
+	{
+		type_t retval = nbl::hlsl::workgroup::inclusive_scan<bin_op_t, ITEMS_PER_WG>::template __call<ScratchProxy<type_t, 0> >( value, prefixSumsAccessor );
+		// we barrier before because we alias the accessors for Binop
+		prefixSumsAccessor.workgroupExecutionAndMemoryBarrier();
+		return retval;
+	}
+};
+
+prefix_sum_t workgroupPrefixSum;
+
+float32_t workgroupBroadcast( NBL_CONST_REF_ARG(float32_t) value, uint32_t index )
+{
+	using accessor_t = ScratchProxy<float32_t, arithmeticSz>;
+	return nbl::hlsl::workgroup::Broadcast<float32_t, accessor_t>( value, broadcastAccessor, index );
+}
+
+// Todo: This spillage calculation is hacky! The lower bound of `_NBL_GLSL_EXT_BLUR_SPILLAGE_LOWER_BOUND_` is just an adhoc
+// thing which happens to work for all images until now but it could very well break for the next image.
+// This:
+// `#define LOCAL_SPILLAGE ITEMS_PER_THREAD`
+// seems to be a safe lower bound, however, it must be inefficient.
+static const uint32_t SPILLAGE_LOWER_BOUND = ( ITEMS_PER_THREAD / 4 );
+
+static const uint32_t IMPL_LOCAL_SPILLAGE = ( ( arithmeticSz - 1 ) / WORKGROUP_SIZE + 1 );
+
+/*
+template<uint32_t T = ITEMS_PER_THREAD>
+static const uint32_t chooseSpillageSize()
+{
+	if( IMPL_LOCAL_SPILLAGE > T )
+	{
+		return T;
+	}
+	else if( IMPL_LOCAL_SPILLAGE < T )
+	{
+		return SPILLAGE_LOWER_BOUND;
+	}
+
+	return IMPL_LOCAL_SPILLAGE;
+}
+*/
+
+static const uint32_t LOCAL_SPILLAGE =
+( IMPL_LOCAL_SPILLAGE > ITEMS_PER_THREAD ) ? ITEMS_PER_THREAD :
+	( ( IMPL_LOCAL_SPILLAGE < ITEMS_PER_THREAD ) ? SPILLAGE_LOWER_BOUND : IMPL_LOCAL_SPILLAGE );
+
+
+uint32_t DataIndex( uint32_t i ) // ?
+{
+	return ( i * gl_WorkGroupSize().x ) + nbl::hlsl::glsl::gl_LocalInvocationIndex();
+}
+
+//TODO: glsl's mod(x, y) = x - y * floor(x/y). Would prolly need to implement a glsl mod
```

Comment:
create a SPV intrinsic for it, you will have to include `unified/glsl.std450.hpp` or something like that (and add it to Nabla builtins - see how the core spirv header was added in CMake)